README.md
[![Latest Stable Version](https://poser.pugx.org/tigerwill90/xsrf-middleware/v/stable)](https://packagist.org/packages/tigerwill90/xsrf-middleware)
[![Latest Unstable Version](https://poser.pugx.org/tigerwill90/xsrf-middleware/v/unstable)](https://packagist.org/packages/tigerwill90/xsrf-middleware)
[![License MIT](https://img.shields.io/badge/License-MIT-brightgreen.svg)](LICENSE.md)
[![Build Status](https://travis-ci.org/tigerwill90/xsrf-middleware.svg?branch=master)](https://travis-ci.org/tigerwill90/xsrf-middleware)
[![codecov](https://codecov.io/gh/tigerwill90/xsrf-middleware/branch/master/graph/badge.svg)](https://codecov.io/gh/tigerwill90/xsrf-middleware)


# PSR-7 & PSR-15 : CSRF Protection alternative for JWT/Branca Authentication token

Csrf protection based on double submit pattern, cookie - JWT/Branca alternative.

It is based on [PSR-7 JWT Authentication Middleware](https://github.com/tuupola/slim-jwt-auth) from
[Tuupola](https://github.com/tuupola). **This middleware is designed to work with
JWT/Branca Authentication method and can be used with any framework using PSR-7 or PSR-15 style middlewares (since v1.1.0). It has been tested with [Slim Framework](https://www.slimframework.com/)**.

This middleware does **not** provide ways to generate Branca/JWT token. However you can find all you
needs for generate token with links bellow.

* [Firebase/php-jwt](https://github.com/firebase/php-jwt)
* [Tuupola/branca-php](https://github.com/tuupola/branca-php)
* [Tuupola/base62](https://github.com/tuupola/base62)

The goal is to protect rest api again [Cross-site request forgery](https://en.wikipedia.org/wiki/Cross-site_request_forgery)
attak, using double submit pattern (stateless).

### How it's work ?

Sometimes you want save your Jwt/Branca token in a http only cookie. Since it's not possible to grab it, your payload content is safe. It's particularly true
for JWT who have no-encrypted payload. BUT, this protection expose your api to CSRF attack.

When a user authenticate to a site

* generate an anti-csrf `token` with pseudorandom value
* generate `JWT` or `Branca` and set one of payload attribute with the previously `token` generated
* send `JWT` or `Branca` to frontend in a `http-only`, `secure` cookie.
* send the previously `token` generated in the response body

When an authenticated api consumer want access to your api, you need to attach the anti-csrf `token` as

* eventually a cookie with unique name
* a header proprieties
* a request body parameter

For all unsafe operation `[POST | PUT | PATCH | DELETE]` to you api, the middleware inspect both `token` and `JWT` or `Branca` in `http-only` cookie to check if value match 
and return [401 status](https://httpstatuses.com/401) if not.

### Dependencies

* [dflydev-fig-cookies](https://github.com/dflydev/dflydev-fig-cookies)
* [tuupola/callable-handler](https://github.com/tuupola/callable-handler)
* [tuupola/http-factory](https://github.com/tuupola/http-factory)
* [rybakit/msgpack](https://packagist.org/packages/rybakit/msgpack)
* php-fig standards


### Install

````
composer require tigerwill90/xsrf-middleware
````

### Usage

Configuration options are passed as an array. There is no mandatory parameter.

```php
$app = new Slim\App

$app->add(new Tigerwill90\Middleware\XsrfProtection([]));
```

When a request is made, the middleware inspect both token and cookie to check if value match. If cookie or token
is not found, the server will respond with `401 Unauthorized`

### Optional parameters
#### Path
The optional `path` parameter allows you to specify which ressources of your api is protected by
the double submit pattern. It can be either a string or an array. You do not need to specify each URL.

Default parameter is `/`
```php
$app = new Slim\App

$app->add(new Tigerwill90\Middleware\XsrfProtection([
    "path" => "/api" /* or ["/api", "/admin"]*/
]));
```

In this example, everything starting with `/api` will be protected.

#### Passthrough

The optional `passthrough` parameter allows you to specify an exceptions to `path` parameter.
It can be either a string or an array.

Default parameter is `null`
```php
$app = new Slim\App

$app->add(new Tigerwill90\Middleware\XsrfProtection([
    "path" => ["/api", "/admin"],
    "passthrough" => "/api/orders"
]));
```

In this example, everything starting with `/api` and `/admin` will be protected, **except** `/api/orders`

#### AntiCsrf 

The optional `anticsrf` parameter allow you to specify the name of your anti-csrf cookie, header or parameter.

Default parameter is `xCsrf`
 ```php
 $app = new Slim\App
 
 $app->add(new Tigerwill90\Middleware\XsrfProtection([
     "path" => ["/api", "/admin"],
     "anticsrf" => "xCsrf"
 ]));
 ```
 
 In this example, if the cookie, header or request parameter "xCsrf" exist, the middleware will compare his value with
 the specified JWT/Branca token `claim` value.
 
 #### Token
 
 According to [PSR-7 JWT Authentication Middleware](https://github.com/tuupola/slim-jwt-auth) documentation, when the token
 is decoded successfully and authentication succees, the contents of decoded token is saved as attribute
 to the `$request`. The optional `token` parameter allows you to specify the attribute name of JWT/Branca token
 that the middleware needs to find in `$request`.
 
 Default parameter is `token`
  ```php
$app = new Slim\App

$app->add(new Tigerwill90\Middleware\XsrfProtection([
  "path" => ["/api", "/admin"],
  "token" => "jwt"
]));
  ```
 
 #### Payload
 
 **Alternatively** you can pass the contents of decoded token in the optional `payload` parameter.
 
 Default value is `null`
 `````php
 $app = new Slim\App
 
$app->add(new Tigerwill90\Middleware\XsrfProtection([
   "path" => ["/api", "/admin"],
   "payload" => $container["decoded"]
]));
 `````
 
 #### Claim
 Beauty of JWT/Branca is that you can pass extra data in the token such roles, rights, etc... Therby, we can
 compare a specified claims with ``httponly`` cookie.
 
 ```php
 [
    "uid" => 1,
    "iat" => "1428819941",
    "exp" => "1744352741",
    "aud" => "www.example.com",
    "roles" => [1,0,1,1,1],
    "xsrf" => "thepseudorandomvaluegeneratedforbothcookieandtoken"
 ]
 
 ```
 
 The optional `claim` parameter allows you to specify the name of the `claim` that the middleware need to find in decoded JWT/Branca token.
 
 Default value is `csrf`
```php
$app = new Slim\App

$app->add(new Tigerwill90\Middleware\XsrfProtection([
  "path" => ["/api", "/admin"],
  "claim" => "xsrf"
]));
```

According to this example, when a request is send to your api, you should have in the header a 
`httponly` cookie and an `authorization` token who have both `thepseudorandomvaluegeneratedforbothcookieandtoken`
setted as value.

#### Logger

The optional `logger` parameter allows you to pass a PSR-3 compatible logger to deal with debugging.

````php
use Monolog\Logger;
use Monolog\Handler\RotatingFileHandler;
use Monolog\Formatter\LineFormatter;

$app = new Slim\App

$logger = new Logger("slim");
$formatter = new LineFormatter(
    "[%datetime%] [%level_name%]: %message% %context%\n",
    null,
    true,
    true
);

$rotating = new RotatingFileHandler(__DIR__ . "/logs/xsrf.log", 0, Logger::DEBUG);
$rotating->setFormatter($formatter);
$logger->pushHandler($rotating);

$app->add(new Tigerwill90\Middleware\XsrfProtection([
  "path" => ["/api", "/admin"],
  "claim" => "xsrf",
  "logger" => $logger
]));
````

In this example we pass an instance of [Logger](https://github.com/projek-xyz/slim-monolog) `$logger` to the middleware.

````
[2017-12-06 01:14:05] [WARNING]: Payload not found in parameter 
[2017-12-06 01:14:05] [DEBUG]: Token and cookie don't match, access denied ! 
````

#### Error

Error is called when access is denied. It receives last error message in arguments.

````php
$app = new Slim\App

$app->add(new Tigerwill90\Middleware\XsrfProtection([
  "path" => ["/api", "/admin"],
  "claim" => "xsrf",
  "error" => function ($response, $arguments) {
       $data["message"] = $arguments["message];
       return $response
                ->withHeader("Content-Type", "application/json")
                ->write(json_encode($data));
  }
]));
````

#### MessagePack

The optional `msgpack` parameter allows you to use the [MessagePack](https://msgpack.org/) serialization format.

Default value is `false`

````php
$app = new Slim\App
 
$app->add(new Tigerwill90\Middleware\XsrfProtection([
   "path" => ["/api", "/admin"],
   "payload" => $container["decoded"]
   "msgpack" => true
]));
````

### Implementation with JWT/Branca Authentication Middleware

Branca/JWT Authentication Middleware need to run before Xsrf Middleware protection.

```php
$container = $app->getContainer();

$container["XsrfProtection"] = function($c) {
    function new \Tigerwill90\Middleware\XsrfProtection([
        "path" => "/api",
        "passthrough" => ["/api/users/signin", "/api/users/token"],
        "anticsrf" => "xCsrf",
        "token" => "jwt",
        "claim" => "xsrf"
    ]);
};

 $container["JwtAuthentication"] = function($c) {
    return new \Slim\Middleware\JwtAuthentication([
        "secure" => true,
        "path" => "/api",
        "passthrough" => ["/api/users/signin", "/api/users/token"],
        "attribute" => "jwt",
        "secret" => getenv("JWT_SECRET")
    ]);
 };
 
 $app->add("XsrfProtection");
 $app->add("JwtAuthentication");
```

### Testing

```
phpunit
```

### License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## [1.3.1-stable](https://github.com/tigerwill90/xsrf-middleware/compare/1.3.0...1.3.1)
##### 2018.02.19
* Bug fix for undefined index when no anti-csrf token is found
* PSR1 & PSR2 coding style

## [1.3.0-stable](https://github.com/tigerwill90/xsrf-middleware/compare/1.2.1...1.3.0)
##### 2018.02.17
Double submit pattern with json alternative is only relevant when Branca/JWT is in a `http-only` cookie. It's particularly
true for JWT who have no-encrypted payload. In http-only cookie, you payload is safe but unsafe again CSRF attack.
* Anti-csrf value is
    * can be found in a anti-csrf cookie (manually setted from client with a different name than original cookie)
    * can be found in header
    * can be found in request parameter
* Jwt/Branca should be always in http-only cookie (it's different for oAuth)
* Refactor some line of code to match with this pattern
* Add unpacking method for MessagePack format

## [1.2.1-stable](https://github.com/tigerwill90/xsrf-middleware/compare/1.2.0...1.2.1)
##### 2018.02.06
* Anti csrf can be find in header
* Cookie option is now anticsrf option
* More test

## 1.2.0-stable
##### 2018.02.15
* Only double submit check for unsafe methods

## 1.1.0
##### 2018.02.13
* Refactor to PHP 7.1
* PSR-7 and PSR-15 support
* Minor improvement

## 1.0.3-release
##### 2017.12.05
* Add support PSR-3 logger
* Add error message
* New sets of tests
* Some code modification
* Stable release

## 1.0.2-beta
##### 2017.11.28
* New tag for packagiste

## 1.0.1-beta
##### 2017.11.28
* New payload parameter
* Passing decoded token in parameter

## 1.0.0-beta
##### 2017.11.20
* Major bug fix
* Support for Branca/JWT Authentication Middleware
* Unit test

tests/bootstrap.php
<?php
/**
 * This file is part of xcsrf-middleware package
 *
 * Copyright (c) 2017 Sylvain Muller
 *
 * Project home : https://github.com/tigerwill90/slim3-xsrf-middleware
 * License : MIT
 */

require __DIR__ . '/../vendor/autoload.php';

tests/XsrfProtectionJwtTest.php
<?php
/**
 * This file is part of xcsrf-middleware package
 *
 * Copyright (c) 2017 Sylvain Muller
 *
 * Project home : https://github.com/tigerwill90/slim3-xsrf-middleware
 * License : MIT
 */

declare(strict_types=1);

namespace Tigerwill90\Middleware;

require __DIR__ . '/../vendor/autoload.php';

use Slim\Http\Request;
use Slim\Http\Response;
use Slim\Http\Uri;
use Slim\Http\Headers;
use Slim\Http\Body;

use Firebase\JWT\JWT;
use Dflydev\FigCookies\Cookie;
use Dflydev\FigCookies\FigRequestCookies;
use Dflydev\FigCookies\SetCookie;

class XsrfProtectionJwtTest extends \PHPUnit_Framework_TestCase {

    private const KEY = "supersecretkeyyoushouldnotcommit";
    private const XSRF = "csrftoken";

    public function requestFactory() : Request {

        $uri = Uri::createFromString('http://dummy.apitest.com/api/signin');
        $headers = new Headers();
        $cookies = [];
        $serverParams = [];
        $body = new Body(fopen('php://temp', 'r+'));
        return new Request('GET', $uri, $headers, $cookies, $serverParams, $body);

     }

     /** @test */
    public function testAllShouldBeWorkWithPayloadAndReturn200() : void {
        $request = $this->requestFactory();
        $response = new Response();

        $payload = [
            "uid" => 1,
            "csrf" => self::XSRF,
            "scope" => [1,0,1,1]
        ];

        $request = FigRequestCookies::set($request, Cookie::create('xCsrf', self::XSRF));

        $jwt = JWT::encode($payload, self::KEY, "HS256");
        $decoded = JWT::decode($jwt, self::KEY, ["HS256", "HS512", "HS384"]);

        $xsrfProtection = new XsrfProtection([
            "payload" => $decoded
        ]);

        $next = function($request, $response) {
            $response->getBody()->write("Foo");
            return $response;
        };


        $response = $xsrfProtection($request,$response, $next);

        $this->assertEquals(200, $response->getStatusCode());
        $this->assertEquals("Foo", $response->getBody());
    }

    /** @test */
    public function testAsArrayShouldMatchAndReturn200() : void {

        $request = $this->requestFactory();
        $response = new Response();

        $payload = [
            "uid" => 1,
            "csrf" => self::XSRF,
            "scope" => [1,0,1,1]
        ];

        $request = FigRequestCookies::set($request, Cookie::create('xCsrf', self::XSRF));

        $jwt = JWT::encode($payload, self::KEY, "HS256");
        $decoded = JWT::decode($jwt, self::KEY, ["HS256", "HS512", "HS384"]);

        $request = $request->withAttribute("token",$decoded);

        $xsrfProtection = new XsrfProtection([]);

        $next = function($request, $response) {
            $response->getBody()->write("Foo");
            return $response;
        };


        $response = $xsrfProtection($request,$response, $next);

        $this->assertEquals(200, $response->getStatusCode());
        $this->assertEquals("Foo", $response->getBody());
    }

    /** @test */
    public function testAsJsonShouldMatchAndReturn200() : void{
        $request = $this->requestFactory();
        $response = new Response();

        $payload = [
            "uid" => 1,
            "csrf" => self::XSRF,
            "scope" => [1,0,1,1]
        ];

        $request = FigRequestCookies::set($request, Cookie::create('xCsrf', self::XSRF));

        $jwt = JWT::encode(json_encode($payload), self::KEY, "HS256");
        $decoded = JWT::decode($jwt, self::KEY, ["HS256", "HS512", "HS384"]);

        $request = $request->withAttribute("token",$decoded);

        $xsrfProtection = new XsrfProtection([]);

        $next = function($request, $response) {
            $response->getBody()->write("Foo");
            return $response;
        };


        $response = $xsrfProtection($request,$response, $next);

        $this->assertEquals(200, $response->getStatusCode());
        $this->assertEquals("Foo", $response->getBody());
    }

    /** @test */
    public function testShouldNotMatchAndReturn401() : void {
        $request = $this->requestFactory();
        $response = new Response();

        $payload = [
            "uid" => 1,
            "csrf" => self::XSRF,
            "scope" => [1,0,1,1]
        ];

        $request = FigRequestCookies::set($request, Cookie::create('xCsrf', "xsrfNotMatch"));

        $jwt = JWT::encode(json_encode($payload), self::KEY, "HS256");
        $decoded = JWT::decode($jwt, self::KEY, ["HS256", "HS512", "HS384"]);

        $request = $request->withAttribute("token",$decoded);

        $xsrfProtection = new XsrfProtection([]);

        $next = function($request, $response) {
            $response->getBody()->write("Foo");
            return $response;
        };


        $response = $xsrfProtection($request,$response, $next);

        $this->assertEquals(401, $response->getStatusCode());
    }

    /** @test */
    public function testShouldNotFindClaimAndReturn401() : void {
        $request = $this->requestFactory();
        $response = new Response();

        $payload = [
            "uid" => 1,
            "noRightClaim" => self::XSRF,
            "scope" => [1,0,1,1]
        ];

        $request = FigRequestCookies::set($request, Cookie::create('xCsrf', "xsrfNotMatch"));

        $jwt = JWT::encode(json_encode($payload), self::KEY, "HS256");
        $decoded = JWT::decode($jwt, self::KEY, ["HS256", "HS512", "HS384"]);

        $request = $request->withAttribute("token",$decoded);

        $xsrfProtection = new XsrfProtection([]);

        $next = function($request, $response) {
            $response->getBody()->write("Foo");
            return $response;
        };


        $response = $xsrfProtection($request,$response, $next);

        $this->assertEquals(401, $response->getStatusCode());
    }

    /** @test */
    public function testShouldNotFindClaimInStringAndReturn401() : void {
        $request = $this->requestFactory();
        $response = new Response();

        $payload = "payloadshouldanassociativearray";

        $request = FigRequestCookies::set($request, Cookie::create('xCsrf', "xsrfNotMatch"));

        $jwt = JWT::encode(json_encode($payload), self::KEY, "HS256");
        $decoded = JWT::decode($jwt, self::KEY, ["HS256", "HS512", "HS384"]);

        $request = $request->withAttribute("token",$decoded);

        $xsrfProtection = new XsrfProtection([]);

        $next = function($request, $response) {
            $response->getBody()->write("Foo");
            return $response;
        };


        $response = $xsrfProtection($request,$response, $next);

        $this->assertEquals(401, $response->getStatusCode());
    }

    /** @test */
    public function testShouldNotFindCookieAndReturn401() : void {
        $request = $this->requestFactory();
        $response = new Response();

        $payload = "payloadshouldanassociativearray";

        $request = FigRequestCookies::set($request, Cookie::create('noRightCookieName', "xsrfNotMatch"));

        $jwt = JWT::encode(json_encode($payload), self::KEY, "HS256");
        $decoded = JWT::decode($jwt, self::KEY, ["HS256", "HS512", "HS384"]);

        $request = $request->withAttribute("token",$decoded);

        $xsrfProtection = new XsrfProtection([]);

        $next = function($request, $response) {
            $response->getBody()->write("Foo");
            return $response;
        };


        $response = $xsrfProtection($request,$response, $next);

        $this->assertEquals(401, $response->getStatusCode());
    }
}

src/XsrfProtection.php
<?php
/**
 * This file is part of xcsrf-middleware package
 *
 * Copyright (c) 2017 Sylvain Muller
 *
 * Project home : https://github.com/tigerwill90/slim3-xsrf-middleware
 * License : MIT
 */

namespace Tigerwill90\Middleware;

use MessagePack\BufferUnpacker;
use Psr\Http\Message\ServerRequestInterface as Request;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;
use Psr\Log\LoggerInterface;
use Psr\Log\LogLevel;
use Dflydev\FigCookies\FigRequestCookies;
use Closure;
use Tuupola\Http\Factory\ResponseFactory;
use Tuupola\Middleware\DoublePassTrait;

final class XsrfProtection implements MiddlewareInterface
{

    use DoublePassTrait;
    
    /**
     * Default options can be overridden
     */
    private $options = [
        "path" => "/",
        "passthrough" => null,
        "payload" => null,
        "anticsrf" => "xCsrf",
        "token" => "token",
        "claim" => "csrf",
        "error" => null,
        "msgpack" => false
    ];

    /**
     * PSR-3 compliant logger
     */
    private $logger;

    /**
     * Last error message
     */
    private $message;

    private const REQUESTMETHOD = ["POST", "PUT", "PATCH", "DELETE"];


    /**
     * Create a new middleware instance
     * @param array
     */
    public function __construct(array $options = [])
    {
        /* Store passed in options overwriting any defaults. */
        $this->setOptions($options);
    }

    /**
     * Process a request in PSR-15 style
     *
     * @param Request $request
     * @param RequestHandlerInterface $handler
     * @return \Psr\Http\Message\ResponseInterface
     */
    public function process(Request $request, RequestHandlerInterface $handler) : Response
    {
        $uri = "/" . $request->getUri()->getPath();
        $uri = preg_replace("#/+#", "/", $uri);
        $requestMethod = $request->getMethod();

        /** If method is safe and mode is AUTH, no need double submit check*/
        if (!\in_array($requestMethod, self::REQUESTMETHOD, true)) {
            $this->log(LogLevel::INFO, "Method " . $requestMethod . " is safe, access granted");
            return $handler->handle($request);
        }

        // if request match with passthrough, no need double submit check
        foreach ((array)$this->options["passthrough"] as $passthrough) {
            $passthrough = rtrim($passthrough, "/");
            if (!!preg_match("@^{$passthrough}(/.*)?$@", $uri)) {
                $this->log(LogLevel::INFO, "Route ignored, access granted");
                return $handler->handle($request);
            }
        }

        // if request match with path, double submit check
        foreach ((array)$this->options["path"] as $path) {
            $path = rtrim($path, "/");
            if (!!preg_match("@^{$path}(/.*)?$@", $uri)) {
                // If anti csrf cannot be found, return 401 Unauthorized
                if (null === $antiCsrfValue = $this->csrfExist($request)) {
                    $response = (new ResponseFactory)->createResponse(401);
                    return $this->error($response, [
                        "message" => $this->message
                    ]);
                }

                // If payload is null and token cannot be found in request, return 401 Unauthorized
                if (!isset($this->options["payload"])) {
                    $message = "Payload not found in options";
                    $this->log(LogLevel::WARNING, $message);
                    if (false === $this->fetchToken($request)) {
                        $response = (new ResponseFactory)->createResponse(401);
                        return $this->error($response, [
                            "message" => $this->message
                        ]);
                    }
                }

                // If claim cannot be found, return 401 Unauthorized
                if (false === $this->fetchClaim()) {
                    $response = (new ResponseFactory)->createResponse(401);
                    return $this->error($response, [
                        "message" => $this->message
                    ]);
                }

                // If antiCsrf don't match with claim, return 401 Unauthorized
                if (false === $this->validateToken($antiCsrfValue)) {
                    $response = (new ResponseFactory)->createResponse(401);
                    return $this->error($response, [
                        "message" => $this->message
                    ]);
                }
            }
        }

        return $handler->handle($request);
    }

    /**
     * Call the error handler if it exists
     *
     * @param Response $response
     * @param array $arguments
     * @return Response
     */
    private function error(Response $response, array $arguments) : Response
    {
        if (\is_callable($this->options["error"])) {
            $handlerResponse = $this->options["error"]($response, $arguments);
            if ($handlerResponse instanceof Response) {
                return $handlerResponse;
            }
        }
        return $response;
    }

    /**
     * Set options from given array (overwrite defaults)
     *
     * @param array $data Array of options.
     * @return self
     */
    private function setOptions(array $data = []) : self
    {
        foreach ($data as $key => $value) {
            $method = "set" . ucfirst($key);
            if (method_exists($this, $method)) {
                \call_user_func(array($this, $method), $value);
            }
        }
        return $this;
    }

    /**
     * Check if anti csrf exist
     *
     * @param Request $request
     * @return null|string
     */
    public function csrfExist(Request $request) : ?string
    {

        /** search anti-csrf in cookie */
        $cookies = FigRequestCookies::get($request, $this->options["anticsrf"]);
        $cookieValue = $cookies->getValue();
        if (isset($cookieValue)) {
            return $cookieValue;
        }

        /** search anti-csrf in header */
        $headers = $request->getHeader($this->options["anticsrf"]);
        $header = $headers[0] ?? null;
        if (isset($header)) {
            return $header;
        }

        /** search anti-csrf in param */
        $params = $request->getParsedBody();
        $param = $params[$this->options["anticsrf"]] ?? null;
        if (isset($param)) {
            return $param;
        }

        /** If no anti-csrf is found, return null */
        $this->message = "Anti csrf not found";
        $this->log(LogLevel::DEBUG, $this->message);
        return null;
    }

    /**
     * Check if payload exist in request attribute
     *
     * @param Request $request
     * @return bool
     */
    public function fetchToken(Request $request) : bool
    {
        $message = "Payload not found in request attribute";
        $decode = $request->getAttribute($this->options["token"]);
        if (!isset($decode)) {
            $this->message = $message;
            $this->log(LogLevel::DEBUG, $message);
            return false;
        }
        $this->options["payload"] = $decode;
        return true;
    }

    /**
     * Check if claim exist
     *
     * @return bool
     */
    public function fetchClaim() : bool
    {
        $decode = $this->transformPayload($this->options["payload"]);
        if (!array_key_exists($this->options["claim"], $decode)) {
            $this->message = "Claim not found in token";
            $this->log(LogLevel::DEBUG, $this->message);
            return false;
        }
        if (empty($decode[$this->options["claim"]])) {
            $this->message = "No random key find in claim";
            $this->log(LogLevel::DEBUG, $this->message);
            return false;
        }

        return true;
    }

    /**
     * Check if anti csrf value match with jwt claim value
     *
     * @param string $antiCsrfValue
     * @return bool
     */
    public function validateToken(string $antiCsrfValue) : bool
    {
        $message = "Token and anti csrf ";
        $decode = $this->transformPayload($this->options["payload"]);
        if ($decode[$this->options["claim"]] === $antiCsrfValue) {
            $this->log(LogLevel::DEBUG, $message . "match, access granted !");
            return true;
        }
        $this->message = $message . "don't match, access denied !";
        $this->log(LogLevel::DEBUG, $this->message);
        return false;
    }

    /**
     * Set a custom path route (overwrite default)
     *
     * @param string[] or string $path
     * @return self
     */
    public function setPath($path) : self
    {
        $this->options["path"] = $path;
        return $this;
    }

    /**
     * get the path route option
     *
     * @return array
     */
    public function getPath() : array
    {
        return (array)$this->options["path"];
    }

    /**
     * Set a custom passthrough route (overwrite default)
     *
     * @param string[] or string $passthrough
     * @return self
     */
    public function setPassthrough($passthrough) : self
    {
        $this->options["passthrough"] = $passthrough;
        return $this;
    }

    /**
     * get the passthrough route option
     *
     * @return string
     */
    public function getPassthrough() : array
    {
        return (array)$this->options["passthrough"];
    }

    /**
     * Set payload given in parameter
     *
     * @param $payload
     * @return $this
     */
    public function setPayload($payload) : self
    {
        $this->options["payload"] = $payload;
        return $this;
    }

    /**
     * Get payload
     *
     * @return array
     */
    public function getPayload() : array
    {
        return $this->transformPayload($this->options["payload"]);
    }

    /**
     * transform payload to valid array
     *
     * @param $payload
     * @return array
     */
    public function transformPayload($payload) : array
    {
        if ($this->options["msgpack"]) {
            $unPacker = new BufferUnpacker();
            $unPacker->reset($payload);
            return (array)$unPacker->unpack();
        }

        if (\is_string($payload)) {
            $isValideDecodedJson = json_decode($payload, true);
            if (json_last_error() === JSON_ERROR_NONE) {
                return (array)$isValideDecodedJson;
            }
        }
        return (array)$this->options["payload"];
    }

    /**
     * Set a custom anti csrf name (overwrite default)
     *
     * @param string $antiCsrf
     * @return self
     */
    public function setAnticsrf(string $antiCsrf) : self
    {
        $this->options["anticsrf"] = $antiCsrf;
        return $this;
    }

    /**
     * get the anti csrf name option
     *
     * @return string
     */
    public function getAnticsrf() : string
    {
        return $this->options["anticsrf"];
    }

    /**
     * Set a custom token name (overwrite default)
     *
     * @param string $token
     * @return self
     */
    public function setToken(string $token) : self
    {
        $this->options["token"] = $token;
        return $this;
    }

    /**
     * get the token name option
     *
     * @return string
     */
    public function getToken() : string
    {
        return $this->options["token"];
    }

    /**
     * Set a custom claim name (overwrite default)
     *
     * @param string $claim
     * @return self
     */
    public function setClaim(string $claim) : self
    {
        $this->options["claim"] = $claim;
        return $this;
    }

    /**
     * Get the claim name option
     *
     * @return string
     */
    public function getClaim() : string
    {
        return $this->options["claim"];
    }

    /**
     * Set a custom PSR-3 compliant logger
     *
     * @param LoggerInterface|null $logger
     * @return $this
     */
    public function setLogger(LoggerInterface $logger = null) : self
    {
        $this->logger = $logger;
        return $this;
    }

    /**
     * Get PSR-3 compliant logger
     *
     * @return mixed
     */
    public function getLogger() : LoggerInterface
    {
        return $this->logger;
    }

    /**
     * Log messages with a PSR-3 compliant logger
     *
     * @param $level
     * @param $message
     * @param array $context
     * @return mixed
     */
    public function log($level, $message, array $context = []) : ?bool
    {
        if ($this->logger) {
            return $this->logger->log($level, $message, $context);
        }
        return false;
    }

    /**
     * Set the last error message
     *
     * @param $message
     * @return $this
     */
    public function setMessage(string $message) : self
    {
        $this->message = $message;
        return $this;
    }

    /**
     * Get the last error message
     *
     * @return mixed
     */
    public function getMessage() : string
    {
        return $this->message;
    }

    /**
     * Set the error handler
     *
     * @param $error
     * @return $this
     */
    public function setError(Closure $error) : self
    {
        $this->options["error"] = $error;
        return $this;
    }

    /**
     * Get the error handler
     *
     * @return mixed
     */
    public function getError() : Closure
    {
        return $this->options["error"];
    }

    /**
     * Get msgpack
     *
     * @param bool $msgPack
     * @return XsrfProtection
     */
    public function setMsgpack(bool $msgPack) : self
    {
        $this->options["msgpack"] = $msgPack;
        return $this;
    }

    /**
     * Set msgpack
     *
     * @return bool
     */
    public function getMsgpack() : bool
    {
        return $this->options["msgpack"];
    }
}
